fortuna/src/command/setup_provider.rs

```diff
@@ -0,0 +1,114 @@
+use {
+    crate::{
+        chain::ethereum::SignablePythContract,
+        command::{
+            register_provider,
+            register_provider::CommitmentMetadata,
+        },
+        config::{
+            Config,
+            RegisterProviderOptions,
+            SetupProviderOptions,
+        },
+        state::{
+            HashChainState,
+            PebbleHashChain,
+        },
+    },
+    anyhow::Result,
+    ethers::{
+        signers::{
+            LocalWallet,
+            Signer,
+        },
+        types::Address,
+    },
+    std::sync::Arc,
+};
+
+/// Setup provider for all the chains.
+/// 1. Register if there was no previous registration.
+/// 2. Re-register if there are no more random numbers to request on the contract.
+/// 3. Re-register if there is a mismatch in generated hash chain.
+/// 4. Update provider fee if there is a mismatch with the fee set on contract.
+/// 4. Update provider uri if there is a mismatch with the uri set on contract.
+pub async fn setup_provider(opts: &SetupProviderOptions) -> Result<()> {
+    let config = Config::load(&opts.config.config)?;
+    let private_key = opts.load_private_key()?;
+    let secret = opts.randomness.load_secret()?;
+    let provider_address = private_key.clone().parse::<LocalWallet>()?.address();
+
+    for (chain_id, chain_config) in &config.chains {
+        // Initialize a Provider to interface with the EVM contract.
+        let contract =
+            Arc::new(SignablePythContract::from_config(&chain_config, &private_key).await?);
+        let provider_info = contract.get_provider_info(provider_address).call().await?;
+
+        let mut register = false;
+
+        // This condition satisfies for both when there is no registration and when there are no
+        // more random numbers left to request
+        if provider_info.end_sequence_number <= provider_info.sequence_number {
+            tracing::info!("endSequenceNumber is lte sequenceNumber.");
```

Comment:
It would be nice to also print these numbers. Also `<=` is more readable than `lte`